### ISSUE TITLE
What about new features like a query batching and subscriptions

### ISSUE THREAD

**Issue by DenisNeustroev** [neelance/graphql-go#15]
https://dev-blog.apollodata.com/new-features-in-graphql-batch-defer-stream-live-and-subscribe-7585d0c28b07

**Comment by neelance** [comment]
Definitely interested! Is there any indication on when this will end up in the official specification or some other spec document?

**Comment by DenisNeustroev** [comment]
No, but Facebook uses these internally. The Apollo (client and server) also supports both features, the official GraphQL server supports only subscriptions.

**Comment by neelance** [comment]
Okay. I'm not against those features, but there are other things I need to take care of first. Effectively this means that those features will either be added when I need them myself or via PR.

**Comment by paralin** [comment]
I'd like to help implement these, ~maybe~

In my project, I do a pipeline where I start out defining the schema in Protobuf, and it gets converted into a variety of different things (TypeScript definitions, GraphQL schema).

Right now I'm looking at an arms race, really:

 - [PlayLyfe](github.com/playlyfe/go-graphql) - got my relatively large project working with some intermediate code-generation and interface manipulations (I generate a function to get the GraphQL schema type name for all of my ProtoBuf schemas). The code base is shit, though. I don't see adding subscriptions to that being a very easy thing.
 - [Go-Graphql](https://github.com/graphql-go/graphql) - massive, seems well made, but pretty inactive and for some reason there is no documentation / any way to convert a schema AST into a Go schema. Supports subscriptions (!)
 - [This repo](https://github.com/neelance/graphql-go) - nicely written, requires some intermediate code generation from me, but otherwise I like it more than the others. The main difficulty here is I have to do [this stuff](https://github.com/neelance/graphql-go/blob/master/example/starwars/starwars.go#L394) which I really hate. I can code-gen most of this, but it just seems like a mess. Some reflection could work better.

Either way, if we could support subscriptions in this repo I think I would probably adopt it for my projects. I'm going to try to write the code-gen stuff in my codebase to make everything work as-is with this repo. If I can do that, I'd love to help implement subscriptions.

**Comment by paralin** [comment]
I've ported all my stuff to run on this library now, nice! Requires a LOT of code generation to cover trivial member field struct resolvers, but it's worth it I think. 

As for subscriptions, where are we on that? I can have a look at what would be required to add them in. Thoughts on what the API could look like @neelance ?

I'm thinking `Exec(` signature is the same, but it just doesn't return until the subscription terminates or the context terminates.

**Comment by neelance** [comment]
What documentation is out there on how subscriptions are supposed to work? How do the updates get pushed to the client? WebSockets?

**Comment by paralin** [comment]
I've been reading the Apollo client stuff on it, but admittedly there doesn't seem to be much. I'm using websockets and I think that's the common way to do it. I send even normal graphql requests over websockets as is right now. 

There's a proposed revision to Graphql to add @defer and @live (I think) to fields. Defer would allow the field to be loaded asynchronously, and the data sent after the initial document is returned. Live would allow updates to that field to be transmitted in real time. 

There's no real info about how canceling / completing these would work.

Supposedly Facebook uses them internally already. 

I've been reading the article in the first message in this thread too - it has most of the info about the proposed features. 

As far as I know subscriptions are done and in the spec, and basically just allow ADD operations to root query fields over time. It seems very limited and I like the defer and live annotations better, but those don't seem to be ready for primetime yet.

**Comment by paralin** [comment]
Facebook seems to use all of these features internally, and just hasn't bothered to put it into the spec / OSS code yet.

Other related issues:

 - https://github.com/apollostack/graphql-server/issues/34
 - https://github.com/facebook/relay/issues/541

Apollo says "we can work on it early march," not sure on relay.

**Comment by paralin** [comment]
In terms of `graphql-go`. Right now we have a single `Execute(` call which returns a single response object. Obviously we need a API change if we want to handle multiple responses.

There also needs to be a mechanism to pair messages coming from a server with a specific query context. In [apollo-client](https://github.com/apollostack/apollo-client/blob/735ee9b49a834e5db65e60a90148f65acd7fd82e/src/transport/networkInterface.ts#L56) this is left up to the network transport, and returns an integer ID for the query context, which can later be released (like a `setTimeout` handle). This is a good approach in my opinion.

I'm already using a similar approach in my WebSocket transport between Apollo and GraphQL-Go for just regular requests - since we don't have the context of a HTTP request, there needs to be an identifier for each request. I think this is best left up to the user / network transport to handle. There are a number of varying approaches to do this depending completely on the transport - if you're using long-lived HTTP requests for streaming requests, for example, such an identifier might be unnecessary as you still have the structure of an HTTP request.

Given that we are leaving this "data channel" stuff up to the user, it then makes sense to have `graphql-go` support a second mode of operation that would allow these kinds of deferred requests. Something like `ExecChannel` which would return a channel of `*Response` objects instead of just a single response. Again, the server implementation can indicate query completion when closing the response channel, and the requester can terminate the long-lived query by canceling the parent context.

Since you could send standard queries alongside complex long-lived queries together through this kind of call, it then makes sense to standardize the server code. Calling resolvers should be the same code regardless of `Channel` or no `Channel`. We can add a parameter to `iExec.exec` for "channel" which, when false, would cause the server to completely ignore directives like `defer` and `live`. 

Right now `exec` returns an `interface{}`. Lists are executed in parallel, spawning a goroutine for each element in the list, remembering the index the result should be in. This code is already super close to what would be necessary to do a `@stream` directive, which returns array elements over time as they are resolved (although index would probably not be necessary in this case). Relevant code for processing things like lists is [here](https://github.com/neelance/graphql-go/blob/master/internal/exec/exec.go#L393). 

I would propose allowing resolvers to return a channel, rather than an array, with an optional argument for a struct with information about the request (is it `@stream` or not?). If we are running a standard `Execute` we can wait until the channel is closed to return the result from that resolver. If we are running an `ExecuteChannel` with the `@stream` set on the field, we can send each result entry in the channel as it comes in over the wire, and not wait for the channel to close to send the initial response.

Doing `@live` is a bit more complex, since we need a way to describe changes to entries we've already returned. I'll leave this for later.

Again, apologies for the essays, but I need to write this somewhere and I figure here's as good a place as any.

**Comment by neelance** [comment]
Is it allowed to use `@live` and `@stream` everywhere? I'd expect that the schema would need to explicitly say where those modifiers are possible. In that case you could expect a different method signature for the resolver.

Imho the features are not yet properly specified. I'd rather wait for Relay 2 to be officially released. Relay 2 is supposed to have those features, so when they release it they will also have to add documentation or a spec and it will also be an official reference implementation, not just something that graphql-js happens to support.

Good luck with any experiments you want to do in a fork, I hope you can make it work for your use case. But please don't be disappointed if we can't bring it into this repo because it is too experimental.

**Comment by paralin** [comment]
That's fine! I know it's experimental.

Placing these specifiers on fields doesn't need a schema change, I think. For stream and defer, it should always work (stream on arrays or channels at least). For live, if we want to be aggressive we can poll the field for changes with a configurable rate. I don't think that's necessary, though. Just treat live like defer if there is no chan returned.

**Comment by paralin** [comment]
@neelance `@defer` is done at https://github.com/paralin/graphql-go/tree/feat-defer

**Comment by paralin** [comment]
@neelance first pass of `@stream` is done.

```
Initial: {"data":{}}
Live: {"data":[null,null,null],"extensions":{"path":["regions"]}}
Live: {"data":{"__typename":"Region","id":"test3"},"extensions":{"path":["regions",2]}}
Live: {"data":{"__typename":"Region","id":"test2"},"extensions":{"path":["regions",1]}}
Live: {"data":{"__typename":"Region","id":"test"},"extensions":{"path":["regions",0]}}
Live: {"data":"test region","extensions":{"path":["regions",0,"name"]}}
Live: {"data":"test region, delayed by 500ms","extensions":{"path":["regions",1,"name"]}}
Live: {"data":"test region, delayed by 1sec","extensions":{"path":["regions",2,"name"]}}
```

Same branch.

**Comment by paralin** [comment]
Arrays can now return channels: `<-chan *ResolverType`

**Comment by paralin** [comment]
@neelance And finally, I have finished `@live`. here are some demos:

 - Nothing but @live on just one sub-field: https://asciinema.org/a/7mzm6ohbkzg1bof0sdl8euyxz
 - Just @stream on the list: https://asciinema.org/a/7pl3z700mf7zhx5rdnhnbx785
 - No new directives (`Exec` compatible) https://asciinema.org/a/1s5f7dm9r7f6l58ix9vc35pna
 - Just `@defer` on the status field: https://asciinema.org/a/8tb0zbkxj81gi0os6t748z2kp

**Comment by bsr203** [comment]
FYI https://github.com/facebook/graphql/pull/267

**Comment by paralin** [comment]
FYI I've moved all of my work from this thread and rewritten everything into http://github.com/rgraphql/magellan

**Comment by mishudark** [comment]
I saw at graphql-go/graphql, and seems like they have a lot of pending PR, also the code has more boilerplate, I think the neelance has a better aproach

**Comment by LeeWong** [comment]
I also think that the neelance has a better aproach. Facebook modern relay is finally out and it supports graphQL subscriptions.  I am wondering if this library would support graphQL subscriptions? Thanks.

**Comment by lpalmes** [comment]
I've been looking at `relay-modern` and they don't have documentation yet for how subscriptions should work, it should be possible to follow apollo idea which i believe it should work with relay out of the box.

**Comment by LeeWong** [comment]
@lpalmes Any update on your latest experience with relay-modern on subscriptions? Thanks.

**Comment by neelance** [comment]
I'm still waiting for an update to the spec.

**Comment by mikeifomin** [comment]
@neelance https://github.com/facebook/graphql/blob/master/rfcs/Subscriptions.md

**Comment by mikeifomin** [comment]
But I guess neelance is waiting updates on this page https://facebook.github.io/graphql/

**Comment by lpalmes** [comment]
@LeeWong i'm sorry i've missed the notification, i'm having way too many github notifications, my bad. I have to try subscriptions in relay-modern and is something that just now has been added to the [docs](https://facebook.github.io/relay/docs/subscriptions.html).

@neelance The spec has been updated and now it has [subscriptions](https://facebook.github.io/graphql/#sec-Subscription) by this [pull request](https://github.com/facebook/graphql/pull/305) and it has been implemented in [graphql-js](https://github.com/graphql/graphql-js/pull/846), to have a reference implementation.

**Comment by LeeWong** [comment]
@lpalmes Thanks. Very excited to see how the relay-modern could work well with this graphql-go, especially in the subscriptions.

**Comment by tonyghita** [comment]
Subscriptions have been merged into the specification: https://facebook.github.io/graphql/#sec-Subscription

I've started reading through... let's do this!

**Comment by tonyghita** [comment]
FWIW I've implemented query batching on top of this library. Works great with the Apollo client!

**Comment by sdemontfort** [comment]
Hey @tonyghita, how did you go about implementing query batching on top of this library? Could you show me an example?

**Comment by nicksrandall** [comment]
@sdemontfort Here is an example of implementing batched queries. https://github.com/nicksrandall/batched-graphql-handler

**Comment by tonyghita** [comment]
@sdemontfort I've been a bit short on time, but I'm planning on making an example repo like @nicksrandall to give a different approach. 

The idea is more or less the same though.

**Comment by ijsnow** [comment]
Hi all! Are you happy with your decision to go with this library rather than https://github.com/graphql-go/graphql? From what I've gathered, this project is cleaner yet less complete in comparison and is more actively maintained. That could either be a result of it being less complete or a good sign of ongoing development. 

I'm just trying to decide on which route I should take when starting to use graphql in a go project I'm starting soon. I asked a similar question over there https://github.com/graphql-go/graphql/issues/196#issuecomment-311431814

Thanks!

**Comment by ijsnow** [comment]
I like the API of this project a lot more. It's so much cleaner. I think I'll go with this one. I went through both star wars examples for the repos and based my decision on the number of questions I had on each and how hard they were to figure out. Although this project's example had a couple hundred more lines of code and no documentation, I could much more easily understand what was going on and trace it through the library.

**Comment by ijsnow** [comment]
I'll do my best to help this repo move forward in the future as well. I may start contributing by writing documentation :P

**Comment by neelance** [comment]
@ijsnow I'm happy to hear that you like the project. As a heads-up: The API will get a major change soon, see #88.

**Comment by mikeifomin** [comment]
No news about the subscription?

**Comment by juliandroid** [comment]
I saw that 6 months ago, some part of the "subscription" support was done, but no tests and probably the most important part is not there yet. Any feedback?

**Comment by mikeifomin** [comment]
@juliandroid did you mean that the package can parse a subscription keyword?
https://github.com/neelance/graphql-go/search?utf8=%E2%9C%93&q=subscription&type=

Does anyone need GraphQL subscriptions right now? We can discuss a implementation.

**Comment by juliandroid** [comment]
@mikeifomin what are the alternatives? If you want to keep track of some property/data and fetch the update?

**Comment by mikeifomin** [comment]
@juliandroid I already use Apollo in my node.js stack.
But one of my project requires the golang-graphql backend and some live data features. 
For now I have implemented the live data in a pure websocket + custom events + json. This cause 
the invention of a wheel at the frontend.
And looking for more convenient and robust solution like the Apollo server for golang.

**Comment by juliandroid** [comment]
Basically, that means Apollo server (handler for graphql-go), plus subscription support. The only thing that I've found so far was github.com/nicksrandall/batched-graphql-handler that handles batching part.

**Comment by paralin** [comment]
Rgraphql should do what you want, and we will be doing stability passes on it in the coming week. It's a very different approach but in the long run I think it will win out. The binary encoding and compression is too good to pass up.

**Comment by juliandroid** [comment]
@paralin Correct me if I'm wrong, but Rgraphql requires graphql-go (the other one) and also Rgraphql from clients site doesn't support iOS and Android as it is the case with Apollo :(

**Comment by AndrewSamokhvalov** [comment]
Hey @neelance! Thanks for the library! Have re-written my API using it, but was initially mislead that it supports subscription, mostly because of `grep` returned some result by keyword `subscription` 😄 . Is it in your near plans to implement it?

**Comment by smithaitufe** [comment]
I need graphql subscriptions as well

**Comment by Kivutar** [comment]
There is an implementation here https://medium.com/functional-foundry/building-graphql-servers-with-subscriptions-in-go-2a60f11dc9f5

**Comment by matiasanaya** [comment]
@pavelnikolov think we can close this one? Subscriptions is now supported and an alternative implementation of query batching has been posted by @nicksrandall.
